Add physical cluster replication tutorials and guides (#17915)
cockroachdb/docs: src/current/v23.2/cutover-replication.md

---

> title: Cut Over from a Primary Cluster to a Standby Cluster

Author: @stevendanna What do you think about the usage of "cut over" throughout these docs? Before Liv left, we discussed that cut over was preferable to "fail over", and "cut over" was generally a known in this context. Do you think it presumes traffic management too, and we should avoid it overall?

> title: Cut Over from a Primary Cluster to a Standby Cluster

Reviewer: I was thinking about that when reading the doc and don't have a great answer.  Personally, I like the word "promote." You "promote a standby to a primary."  But "cut over" has the advantage of it also being the words we used in some error messages, UI elements, and log messages.

As long as we make it clear in the doc that the user is responsible for ensuring traffic goes to the new cluster and that writes that go to the old cluster may be lost (which I think you do mention below). Then perhaps we don't have to get it perfect in this version.

In 24.1 when we have a way to go back quickly, I think we'll be forced to decide on terminology since then we'll need to decide if it is "failover and failback" or "cutover and cutback.

Which is a long way to say, perhaps stick with your current wording for now.

> title: Cut Over from a Primary Cluster to a Standby Cluster

Author: OK, I understand where you're coming from. I have left as "Cutover / to cut over" for now. I honestly prefer "promote" from a language pov. However, if v24.1 is going to include a way to return quickly, and we'll be picking from fail/cut anyway... maybe this should stay.

I have made the note more of a warning re the user having to switch traffic manually.

summary: A guide to complete physical cluster replication and cut over from a primary to a standby cluster.
toc: true
docs_area: manage
---

{{site.data.alerts.callout_info}}
{% include feature-phases/preview.md %}
{{site.data.alerts.end}}

> 

Author: Line 23, I have referred to the cut back process and that a full reverse process is required.

The cutover is a two-step process on the standby cluster:

1. [Initiating the cutover](#step-1-initiating-the-cutover).
1. [Completing the cutover](#step-2-completing-the-cutover).

{{site.data.alerts.callout_info}}
Initiating a cutover is a manual process that makes the standby cluster ready to accept SQL connections. Once the cutover is complete, you will need to ensure that a load balancer {% comment %}or other infra{% endcomment %} redirects application traffic to the standby cluster.
{{site.data.alerts.end}}

{% comment %}This page describes how to cut over from the primary to the standby cluster, for information on how replication works, refer to:{% endcomment %}

{% include {{ page.version.version }}/physical-replication/reference-links-replication.md %}

## Step 1. Initiating the cutover

> You can initiate a cutover to the standby cluster in a couple of different ways. Refer to the following sections for steps:

Author: I have updated this to refer to the difference in specifying the time, hopefully this works.

- [`LATEST`](#cut-over-to-the-most-recent-replicated-time): The most recent replicated timestamp.
- [Point-in-time](#cut-over-to-a-point-in-time):
    - Past: A timestamp in the past that is within the cutover window. {% comment %}Add link to technical overview (replicated timestamp + cutover window){% endcomment %}
    - Future: A timestamp in the future in order to plan a cutover.

> ### Cut over to the most recent replicated time

Author: Done, included your reminder here and a code block for `WITH REPLICATION STATUS`

docs_area: manage
---

{{site.data.alerts.callout_info}}
{% include feature-phases/preview.md %}
{{site.data.alerts.end}}

{% include_cached new-in.html version="v23.2" %} Physical cluster replication allows you to cut over from the active primary cluster to the passive standby cluster that has ingested replicated data. When you complete the replication, it will stop the stream of new data, reset the standby virtual cluster to a point in time where all ingested data is consistent, and then mark the standby virtual cluster as ready to accept traffic.

The cutover is a two-step process on the standby cluster:

1. [Initiating the cutover](#step-1-initiate-the-cutover).
1. [Completing the cutover](#step-2-complete-the-cutover).

{{site.data.alerts.callout_danger}}
Initiating a cutover is a manual process that makes the standby cluster ready to accept SQL connections. However, the cutover process does **not** automatically redirect traffic to the standby cluster. Once the cutover is complete, you must redirect application traffic to the standby (new) cluster. If you do not manually redirect traffic, writes to the primary (original) cluster may be lost.
{{site.data.alerts.end}}

> After a cutover, you may want to _cut back_ to the original primary cluster. That is, set up the original primary cluster to once again accept application traffic. This requires you to undertake another a full replication stream in the opposite direction from standby to primary. For more detail, refer to [Cut back to the primary cluster](#cut-back-to-the-primary-cluster).

Author: Changed!

> {% comment %}This page describes how to cut over from the primary to the standby cluster, for information on how replication works, refer to:{% endcomment %}

Author: I have removed this for now. Going to add back in when I have merged this PR and then work on getting the overviews prepped for merging.

{% include {{ page.version.version }}/physical-replication/reference-links-replication.md %}

## Step 1. Initiate the cutover

> To initiate a cutover to the standby cluster, there are different ways of specifying the point in time for the standby's promotion. Refer to the following sections for steps:

Author: I have added an additional sentence here, keeping "promotion" but defining it.

---+--------------------+--------------------+--------------+--------------------+----------------------------------------------------------------------------------------------------------------------+--------------------+-------------------------------+-------------------------------+---------------
3  | standbyapplication | replicating        | none         | application        | postgresql://{user}:redacted@{hostname}:26257/?options=-ccluster%3Dsystem&sslmode=verify-full&sslrootcert=redacted   | 899090689449132033 | 2023-09-11 22:29:35.085548+00 | 2023-09-11 16:51:43.612846+00 |     NULL
(1 row)
~~~

Specify a timestamp:

{% include_cached copy-clipboard.html %}
~~~ sql
ALTER VIRTUAL CLUSTER standbyapplication COMPLETE REPLICATION TO SYSTEM TIME '-1h';
~~~

Refer to [Using different timestamp formats]({% link {{ page.version.version }}/as-of-system-time.md %}#using-different-timestamp-formats) for more information.

Similarly, to cut over to a specific time in the future:

{% include_cached copy-clipboard.html %}
~~~ sql
ALTER VIRTUAL CLUSTER standbyapplication COMPLETE REPLICATION TO SYSTEM TIME '+5h';
~~~

> A future cutover will proceed once the replicated data has reached the specified time.

Author: So, this is actually covered in the next step. I did try and a sentence here to be like "look at the next step", but it felt a little awkward. Left as is, but going to re-examine this when I work on publishing the SQL reference pages.

> ## Step 2. Completing the cutover

Reviewer 2: ```suggestion
## Step 2. Complete the cutover
```

Author: Chopped that gerund, did the other step too.

    ~~~
    ~~~
      id |        name         |         data_state          | service_mode | source_tenant_name |                                                source_cluster_uri                                                 | replication_job_id |       replicated_time        |         retained_time         |          cutover_time
    -----+---------------------+-----------------------------+--------------+--------------------+-------------------------------------------------------------------------------------------------------------------+--------------------+------------------------------+-------------------------------+---------------------------------
      4  | standbyapplication  | replication pending cutover | none         | application        | postgresql://user:redacted@3ip:26257/?options=-ccluster%3Dsystem&sslmode=verify-full&sslrootcert=redacted         | 903895265809498113 | 2023-09-28 17:41:18.03092+00 | 2023-09-28 16:09:04.327473+00 | 1695922878030920020.0000000000
    (1 row)
    ~~~

    Refer to [Physical Cluster Replication Monitoring]({% link {{ page.version.version }}/physical-cluster-replication-monitoring.md %}) for more  detail.

> 1. Once complete, bring the standby's virtual cluster online with:

Author: I linked to the responses and data state tables as well in this sentence. I think once the SQL ref pages are out, they might also help.

> 1. Once complete, start the standby's virtual cluster with:

Reviewer 2: ```suggestion
1. Once complete, bring the standby's virtual cluster online with:
```

Author: done

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    ALTER VIRTUAL CLUSTER standbyapplication START SERVICE SHARED;
    ~~~

    ~~~
      id |        name         |     data_state     | service_mode
    -----+---------------------+--------------------+---------------
      1  | system              | ready              | shared
      2  | template            | ready              | none
      3  | standbyapplication  | ready              | shared
    (3 rows)
    ~~~

1. To make the standby's virtual cluster the default for connection strings. Set the following [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}):

> 

Reviewer 2: It is worth pointing out that PCR is not enabled on the newly-cut-over now-primary cluster. You'll need to start back at the beginning with a new standby cluster to re-enable it. I guess you do discuss it in the section below but it might be worth linking to there from here.

Author: I clarified this a little bit at the end of the Complete Cutover step, hopefully that captures what you were thinking.

>     {% include_cached copy-clipboard.html %}

Author: Added the cluster setting here, is the brief intro sentence OK?

Reviewer: Yup, looks good.

    ~~~ sql
    SET CLUSTER SETTING server.controller.default_target_cluster='standbyapplication'
    ~~~

> At this point, the primary and standby clusters are entirely independent. You will need to use your own orchestration to direct traffic to the standby.

Reviewer 2: This is very vague. I see why, but saying "orchestration" makes you think of Kubernetes, which is not what this is about. I think you mean something like network load balancing. It's not clear to me if the best practice is to nuke and pave the previous primary since we don't support an Active-Active setup, which it seems like this is likely to lead to.

Author: I changed to load balancers here. I thought this was vague too — and yes, I guess we can't really go into detail here. I wonder if we can broaden this as we have more customers using it

    (3 rows)
    ~~~

1. To make the standby's virtual cluster the default for connection strings, set the following [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}):

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET CLUSTER SETTING server.controller.default_target_cluster='standbyapplication'
    ~~~

> At this point, the primary and standby clusters are entirely independent. You will need to use your own network load balancers to direct application traffic to the standby (now primary). To enable physical cluster replication again, from the new primary to the original primary (or a completely different cluster), refer to [Cut back to the primary cluster](#cut-back-to-the-primary-cluster).

Reviewer 2: ```suggestion
At this point, the primary and standby clusters are entirely independent. You will need to use your own network load balancers, DNS servers, or other network configuration to direct application traffic to the standby (now primary). To enable physical cluster replication again, from the new primary to the original primary (or a completely different cluster), refer to [Cut back to the primary cluster](#cut-back-to-the-primary-cluster).
```

Author: done!
